src/interface_index.rs

```diff
@@ -0,0 +1,391 @@
+/// and not plain text.
+const MARKER: &str = "INSERT INTERFACES HERE";
+
+impl Processor {
+    pub fn new() -> Self {
+        Processor {
+            interfaces: BTreeMap::new(),
+            marker_nodes: Vec::new(),
+        }
+    }
+
+    pub fn visit(&mut self, node: &Handle) {
+        const ID: QualName = QualName {
+            prefix: None,
+            ns: ns!(),
+            local: local_name!("id"),
+        };
+        // We're looking for <code class="idl"> inside a <pre>, to find
+        // potential interfaces defined there.
+        //
+        // One surprise here -- there is an "interface Example" that is not defined
+        // according to Wattsi. It yells about this not being defined, and the
+        // prior Perl preprocessing actually requires the <pre> have no
+        // attributes.
+        if node.is_html_element(&local_name!("code"))
+            && node.has_class("idl")
+            && node.parent_node().map_or(false, |p| {
+                p.is_html_element(&local_name!("pre")) && !p.has_class("extract")
+            })
+        {
+            let borrowed_children = node.children.borrow();
+            for window in borrowed_children.windows(2) {
+                let is_partial = match window[0].node_text() {
+                    Some(a) if a.ends_with("partial interface ") => true,
+                    Some(a) if a.ends_with("interface ") => false,
+                    _ => continue,
+                };
+                // These definitions must appear as a <span>, <dfn> or <a> element.
+                if !window[1].is_html_element(&local_name!("span"))
+                    && !window[1].is_html_element(&local_name!("dfn"))
+                    && !window[1].is_html_element(&local_name!("a"))
+                {
+                    continue;
+                }
+                let name = window[1].text_content();
+                let info = self.interfaces.entry(name).or_default();
+                if is_partial {
+                    if let Some(id) = window[1].get_attribute(&ID) {
+                        info.partials.push(id);
+                    } else {
+                        info.has_partial_with_no_id = true;
+                    }
+                } else {
+                    info.seen += 1;
+                }
+            }
+        }
+
+        if node.node_text().map_or(false, |t| t.contains(MARKER)) {
+            self.marker_nodes.push(node.clone());
+        }
+    }
+
+    pub fn apply(self) -> io::Result<()> {
+        // It is likely an author error to not include anywhere to insert an
+        // interface index. More than one is supported, mainly because it's no
+        // more work than enforcing that just one exists.
+        if self.marker_nodes.is_empty() {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                format!("Marker {MARKER:?} not found."),
+            ));
+        }
+        for marker in self.marker_nodes {
```

Comment:
Can there ever actually be multiple of these? Current `source` seems to only ever have one, so I'm wondering if we should interpret multiple as an error.